Source/AwsCommonRuntimeKit/crt/CBOR.swift

```diff
@@ -151,218 +152,292 @@ public class CBORDecoder {
+        }
+    }
+
+    private func decodeDefiniteArray() throws -> CBORType {
+        var length: UInt64 = 0
+        guard aws_cbor_decoder_pop_next_array_start(self.rawValue, &length) == AWS_OP_SUCCESS else {
+            throw CommonRunTimeError.crtError(.makeFromLastError())
+        }
+
+        var array: [CBORType] = []
+        for _ in 0..<length {
+            array.append(try popNext())
+        }
+        return .array(array)
+    }
+
+    private func decodeDefiniteMap() throws -> CBORType {
+        var out_value: UInt64 = 0
+        guard
+            aws_cbor_decoder_pop_next_map_start(self.rawValue, &out_value)
+                == AWS_OP_SUCCESS
+        else {
+            throw CommonRunTimeError.crtError(.makeFromLastError())
+        }
+        var map: [String: CBORType] = [:]
+        for _ in 0..<out_value {
+            let key = try popNext()
+            if case .text(let key) = key {
+                map[key] = try popNext()
             } else {
                 throw CommonRunTimeError.crtError(CRTError(code: AWS_ERROR_CBOR_UNEXPECTED_TYPE.rawValue))
             }
-        case AWS_CBOR_TYPE_ARRAY_START:
-            var out_value: UInt64 = 0
-            guard
-                aws_cbor_decoder_pop_next_array_start(self.rawValue, &out_value)
-                    == AWS_OP_SUCCESS
-            else {
+        }
+        return .map(map)
+    }
+
+    // Decoding helper methods for break and indefinite types
+    private func decodeBreak() throws -> CBORType {
+        // This should only be called inside indefinite decoding
+        guard aws_cbor_decoder_consume_next_single_element(self.rawValue) == AWS_OP_SUCCESS else {
+            throw CommonRunTimeError.crtError(.makeFromLastError())
+        }
+        return .indef_break
+    }
+
+    private func decodeIndefiniteArray() throws -> CBORType {
+        guard aws_cbor_decoder_consume_next_single_element(self.rawValue) == AWS_OP_SUCCESS else {
+            throw CommonRunTimeError.crtError(.makeFromLastError())
+        }
+        var array: [CBORType] = []
+        while true {
+            var cbor_type: aws_cbor_type = AWS_CBOR_TYPE_UNKNOWN
+            guard aws_cbor_decoder_peek_type(self.rawValue, &cbor_type) == AWS_OP_SUCCESS else {
                 throw CommonRunTimeError.crtError(.makeFromLastError())
             }
-            var array: [CBORType] = []
-            for _ in 0..<out_value {
-                array.append(try popNext())
+
+            if cbor_type == AWS_CBOR_TYPE_BREAK {
+                _ = try decodeBreak()
+                break
             }
-            return .array(array)
-        case AWS_CBOR_TYPE_MAP_START:
-            var out_value: UInt64 = 0
-            guard
-                aws_cbor_decoder_pop_next_map_start(self.rawValue, &out_value)
-                    == AWS_OP_SUCCESS
-            else {
+            array.append(try popNext())
```

Comment:
You can just do something like 
let cborType = try popNext(); 
if cborType == BREAK 
      break 
else 
      array.append(cborType)
      
      
 Same for other indef types.